src/java.base/share/classes/java/lang/Thread.java

```diff
@@ -1729,18 +1729,23 @@ public final void stop() {
      * @revised 6.0, 14
      */
     public void interrupt() {
-        if (currentThread() != this) {
+        if (this != Thread.currentThread()) {
             checkAccess();
         }
 
+        // Setting the interrupt status must be done before reading nioBlocker.
         synchronized (interruptLock) {
             interrupted = true;
             interrupt0();  // inform VM of interrupt
+        }
 
-            // thread may be blocked in an I/O operation
-            Interruptible b = nioBlocker;
-            if (b != null) {
-                b.interrupt(this);
+        // thread may be blocked in an I/O operation
+        if (this != Thread.currentThread()) {
+            synchronized (interruptLock) {
+                Interruptible b = nioBlocker;
+                if (b != null) {
+                    b.interrupt(this);
+                }
             }
```

Comment:
I think we want just one `synchronized` block (we don't want to drop the monitor before considering the nioBlocker).
```java
        // Setting the interrupt status must be done before reading nioBlocker.
        synchronized (interruptLock) {
            interrupted = true;
            interrupt0();  // inform VM of interrupt

            // thread may be blocked in an I/O operation
            if (this != Thread.currentThread()) {
                Interruptible b = nioBlocker;
                if (b != null) {
                    b.interrupt(this);
                }
            }
        }
```

Reply:
What's the risk to drop the monitor before considering the `nioBlocker`?
`synchronized (interruptLock)` for `nioBlocker.interrupt(this)` is to ensure the `nioBlocker` is the one that the current thread is `blockedOn(Interruptible b)`. I don't see it will be affected by `interrupted` or `interrupt0()` of `this` thread.
The current change saves `if (this != Thread.currentThread())` within the `synchronized` block.

Reply:
The old code held the lock across for the same sequence I'm proposing now. The only change we want is the added check whether the thread is interrupted by a distinct thread. If the lock is released in between, it's possible that the interrupted state can change.

Reply:
Moved `if (this != Thread.currentThread())` into the `synchronized` block.